### ISSUE TITLE
THNN doc

### ISSUE THREAD

**Issue by nicholas-leonard** [torch/nn#559]
Would be nice.

**Comment by hughperkins** [comment]
:+1:

**Comment by andreaskoepf** [comment]
You are right, we need to provide some documentation about the THNN changes.

Beside general explanations I think a 'How to extend torch/nn' article that lists the steps required to add a layer or criterion would lower the barrier for new authors to extend torch, e.g. that a contribution should consists of lua, C, cuda, unit-test and documentation. We could also discuss a the frequently appearing scenario that people want add torch functionality in private repositories, e.g. @j-wilson has prepared a [generalization of the THNN call mechanism](https://github.com/torch/nn/compare/master...j-wilson:master)...

**Comment by andreaskoepf** [comment]
A very first version of [THNN doc](https://github.com/torch/nn/tree/master/lib/THNN) is now available thanks to @apaszke.

**Comment by nicholas-leonard** [comment]
@andreaskoepf Its a good start. Seems like it will require a lot of work to general the API doc.

**Comment by hughperkins** [comment]
> Seems like it will require a lot of work to general the API doc

I cant tell whether you are saying you mean it's incomplete, and needs someone to work on it, or whether you're saying it's superfluous, since we can just read the sourcecode to get the signatures anyway?

**Comment by nicholas-leonard** [comment]
@hughperkins I mean this will require a lot of work to complete : https://github.com/torch/nn/blob/master/lib/THNN/doc/api_reference.md

**Comment by j-wilson** [comment]
@soumith @andreaskoepf @apaszke Regarding doc for API, are people still waiting to see whether or not THNN will adopt the [torchAPI](https://github.com/j-wilson/torchAPI) stuff I started working on. was a decision ever made regarding this?

**Comment by hughperkins** [comment]
Ah... In my opinion, I would think that the source-code is plausibly sufficient as an API reference?  However, if we do need a doc for the API reference, I've dabbled a bit with [doxygen](www.stack.nl/~dimitri/doxygen/) in the past, and it seems to give ok-ish results? http://deepcl.hughperkins.com/4.x.x/html/annotated.html

**Comment by apaszke** [comment]
I don't think that the source code is sufficient everywhere. If it's a simple criterion then a quick lookup might be enough, but there are modules that aren't so straightforward. It's also much faster to browse a human readable document.

Also, I don't think we really need doxygen here. It's just a simple functional API. I have nearly finished a script that will regenerate api_reference.md based on the comments from THNN.h, so it will be easy to maintain it in the future.

**Comment by hughperkins** [comment]
> I have nearly finished a script that will regenerate api_reference.md based on the comments from THNN.h, so it will be easy to maintain it in the future.

Ah, that sounds like an excellent plan.  On the one hand, doxygen is really heavy, and on the other hand it probalby wouldnt work without some serious tweaking.  I think your plan sounds really good to me :-)

**Comment by apaszke** [comment]
I've submitted the script in #668.

**Comment by nicholas-leonard** [comment]
Is there an example out there on how to extend THNN from other packages? nnx is still using the old way.

**Comment by apaszke** [comment]
No, I didn't write anything about it yet, because it's still undecided if @j-wilson's changes will be included. Depending on what's going to happen, extending THNN will look a bit different, so I decided to wait for now.

**Comment by j-wilson** [comment]
@apaszke  @nicholas-leonard  @soumith So as to be clear, I'm more than happy to accommodate whatever people feel is appropriate w.r.t. the API stuff. Generally speaking, I'm primarily focused on facilitating the development of C-based Torch7 libraries, which will hopefully help enhance the community as a whole.

TLDR: if everyone uses the same interface to link C and Torch libraries, the overheads for developing new libraries will decrease.
